File: examples/hybrid-notes/ext/ts/notes.ts
// The bespoke half of the hybrid SDK: the two operations the service does not
// serve over HTTP. The generated serde file imports these relative to itself,
// so drop the ext/ directory next to the generated SDK files.
//
// The store here is a deterministic stand-in for whatever the real thing is
// (a legacy library, a cache, a crypto module). It is written to be exercised
// by the conformance vectors, so every branch is reachable from an input, and
// it agrees with the Go implementation case for case.

import type { Outcome } from "@tono/ext-runtime-ts";

import type { Settings } from "../../notes_serde";
import { type Note, type NoteRef, OverloadedError } from "../../notes";

// storeSave is the typed form: it speaks the operation's own types, so it owns
// the mapping and reports a declared failure as the declared value. Throwing
// OverloadedError crosses the generated boundary untouched, retryable() and
// all; throwing anything else becomes a ContractError naming the operation.
export async function storeSave(settings: Settings, input: Note): Promise<Note> {
  switch (input.id) {
    case "overload":
      throw new OverloadedError(
        { message: "the store is shedding load" },
        JSON.stringify({ message: "the store is shedding load" }),
      );
    case "boom":
      throw new Error("the store panicked");
  }
  return { id: input.id, body: input.body, modified: `saved-by-${settings.apiToken}` };
}

// storeArchive is the raw form: it returns the payload the generated glue
// integrates. On success the body is decoded strictly into Note; on failure the
// code is matched against the operation's declared error codes. Neither side
// writes mapping code.
export async function storeArchive(settings: Settings, payload: string): Promise<Outcome> {
  const ref = JSON.parse(payload) as NoteRef;
  switch (ref.id) {
    case "overload":
      return failure("overloaded", '{"message":"the store is shedding load"}');
    case "unknown-failure":
      // No declared error carries this code, so the glue falls back to the
      // generic API error rather than guessing.
      return failure("quota_exceeded", '{"message":"out of quota"}');
    case "corrupt":
      // A success payload missing a required member: the glue rejects it with
      // the path of the member the store omitted.
      return { success: true, code: "", body: '{"id":"corrupt"}' };
    case "boom":
      throw new Error("the store panicked");
  }
  const body = JSON.stringify({
    id: ref.id,
    body: "archived",
    updated_at: `archived-by-${settings.apiToken}`,
  });
  return { success: true, code: "", body };
}

function failure(code: string, body: string): Outcome {
  return { success: false, code, body };
}

File: examples/hybrid-notes/ext/go/notes.go
// The bespoke half of the hybrid SDK: the two operations the service does not
// serve over HTTP. Go calls a bound symbol unqualified, from inside the
// generated package, so this file is dropped next to the generated sources
// (the binding path is a "put it here" instruction, not an import path).
//
// The store here is a deterministic stand-in for whatever the real thing is
// (a legacy library, a cache, a crypto module). It is written to be exercised
// by the conformance vectors, so every branch is reachable from an input.

package notes

import (
	"context"
	"encoding/json"
	"errors"

	tonoext "github.com/tono-lang/tono/runtimes/ext-go"
)

// StoreSave is the typed form: it speaks the operation's own types, so it owns
// the mapping and reports a declared failure as the declared value. Returning
// *Overloaded crosses the generated boundary untouched, Retryable() and all;
// returning anything else becomes a ContractError naming the operation.
func StoreSave(ctx context.Context, s *Settings, input Note) (Note, error) {
	switch input.ID {
	case "overload":
		return Note{}, &Overloaded{Message: "the store is shedding load"}
	case "boom":
		return Note{}, errors.New("the store panicked")
	}
	return Note{ID: input.ID, Body: input.Body, Modified: "saved-by-" + s.APIToken}, nil
}

// StoreArchive is the raw form: it returns the payload the generated glue
// integrates. On success the body is decoded strictly into Note; on failure the
// code is matched against the operation's declared error codes. Neither side
// writes mapping code.
func StoreArchive(ctx context.Context, s *Settings, payload []byte) (tonoext.Outcome, error) {
	var ref NoteRef
	if err := json.Unmarshal(payload, &ref); err != nil {
		return tonoext.Outcome{}, err
	}
	switch ref.ID {
	case "overload":
		return failure("overloaded", `{"message":"the store is shedding load"}`), nil
	case "unknown-failure":
		// No declared error carries this code, so the glue falls back to the
		// generic API error rather than guessing.
		return failure("quota_exceeded", `{"message":"out of quota"}`), nil
	case "corrupt":
		// A success payload missing a required member: the glue rejects it with
		// the path of the member the store omitted.
		return tonoext.Outcome{Success: true, Body: []byte(`{"id":"corrupt"}`)}, nil
	case "boom":
		return tonoext.Outcome{}, errors.New("the store panicked")
	}
	body, err := json.Marshal(Note{
		ID:       ref.ID,
		Body:     "archived",
		Modified: "archived-by-" + s.APIToken,
	})
	if err != nil {
		return tonoext.Outcome{}, err
	}
	return tonoext.Outcome{Success: true, Body: body}, nil
}

func failure(code string, body string) tonoext.Outcome {
	return tonoext.Outcome{Success: false, Code: code, Body: []byte(body)}
}
